From d7bc6b086027355986744d2c1cb95d4edbaf25cb Mon Sep 17 00:00:00 2001
From: Pradyot Bhattacharyya <pbhattacharyya@habana.ai>
Date: Tue, 03 Jun 2025 14:51:42 +0300
Subject: [PATCH] [FSW-9705] fs1: implement gt boot sequence

This adds changes to the gpu ipm driver to support the poll
mechanism required to implement gt init sequence.

Change-Id: If8a0b28701eb574a83a77f263a9c072429b292ef
Signed-off-by: Pradyot Bhattacharyya <pbhattacharyya@habana.ai>
---

diff --git a/drivers/ipm/ipm_gpu_bridge.c b/drivers/ipm/ipm_gpu_bridge.c
index d7b53b0..a5ce798 100644
--- a/drivers/ipm/ipm_gpu_bridge.c
+++ b/drivers/ipm/ipm_gpu_bridge.c
@@ -14,18 +14,22 @@
 
 #define LOG_MODULE_NAME			ipm_gpu_bridge
 LOG_MODULE_REGISTER(LOG_MODULE_NAME, CONFIG_IPM_LOG_LEVEL);
-
+#define INTR_CLR_WAIT_MS	2
+#define TOTAL_TIMEOUT_MS	500
+#define INTERRUPT_DATA_SZ	9
+#define INTR_MSG_HDR_SZ		2
+#define FLIT_SIZE_B		sizeof(uint32_t)
 #define GPU_BRIDGE_NUM_FLITS_MASK	0x1F
 #define GPU_BRIDGE_MAX_NUM_OF_DATA	36 /* 36 bytes */
 
 struct gpu_bridge_cfg_regs {
-	uint32_t interrupt_out_data[9]; /* offset 0x0 */
-	uint32_t interrupt_out_cfg;	/* offset 0x24 */
-	uint32_t interrupt_out_valid;	/* offset 0x28 */
-	uint32_t interrupt_in_data[9];	/* offset 0x2C */
-	uint32_t interrupt_in_cfg;	/* offset 0x50 */
-	uint32_t interrupt_in_valid;	/* offset 0x54 */
-	uint32_t i2a_cfg;		/* offset 0x58 */
+	uint32_t interrupt_out_data[INTERRUPT_DATA_SZ]; /* offset 0x0 */
+	uint32_t interrupt_out_cfg;			/* offset 0x24 */
+	uint32_t interrupt_out_valid;			/* offset 0x28 */
+	uint32_t interrupt_in_data[INTERRUPT_DATA_SZ];	/* offset 0x2C */
+	uint32_t interrupt_in_cfg;			/* offset 0x50 */
+	uint32_t interrupt_in_valid;			/* offset 0x54 */
+	uint32_t i2a_cfg;				/* offset 0x58 */
 };
 
 struct gpu_bridge_intr_regs {
@@ -109,6 +113,10 @@
 {
 	const struct ipm_gpu_bridge_config *config = DEV_CFG(dev);
 	struct gpu_bridge_cfg_regs *cfg_base = (struct gpu_bridge_cfg_regs *)(config->cfg_base);
+	int data_size;
+	uint32_t flitnum;
+	int timeout = 0;
+	int i;
 
 	/*
 	 * Send API makes sure only 1 request can be send at given time.
@@ -120,13 +128,19 @@
 		return -EINVAL;
 	}
 
-	if (size > GPU_BRIDGE_MAX_NUM_OF_DATA) {
+	/*
+	 * data_size includes 32 bits for flitnumber. So subtract
+	 * that to get the size of actual data to send to mailbox
+	 */
+	data_size = size - FLIT_SIZE_B;
+	if (data_size > GPU_BRIDGE_MAX_NUM_OF_DATA) {
 		LOG_ERR("message size %d is greater than supported %d",
-			size, GPU_BRIDGE_MAX_NUM_OF_DATA);
+			data_size, GPU_BRIDGE_MAX_NUM_OF_DATA);
 			return -ENOMEM;
 	}
 
-	if (size % 4) {
+	flitnum = *((uint32_t *)((uint8_t *)data + data_size));
+	if (data_size % 4) {
 		LOG_ERR("size is not 4 bytes aligned");
 		return -EINVAL;
 	}
@@ -141,11 +155,11 @@
 		return -EBUSY;
 	}
 
-	/* Copy data to buffer */
-	memcpy((void *) &cfg_base->interrupt_in_data, data, size);
+	for (i = 0; i < INTERRUPT_DATA_SZ; i++)
+		sys_write32(*((uint32_t *)data + i), (mem_addr_t)&cfg_base->interrupt_in_data[i]);
 
 	/* update number of flits in dword TODO: check what is NP bit in cfg (non posted?) */
-	sys_write32((size >> 2), (mem_addr_t) &cfg_base->interrupt_in_cfg);
+	sys_write32((data_size >> 2) * flitnum, (mem_addr_t) &cfg_base->interrupt_in_cfg);
 
 	/* Raise interrupt */
 	sys_write32(0x1, (mem_addr_t) &cfg_base->interrupt_in_valid);
@@ -154,6 +168,13 @@
 	 * TODO: check whether driver needs to wait for clear of interrupt_in_valid
 	 * indicating bridge processed the message.
 	 */
+	if (wait) {
+		while (sys_read32((mem_addr_t)&cfg_base->interrupt_in_valid)) {
+			k_msleep(INTR_CLR_WAIT_MS);
+			if (timeout++ > TOTAL_TIMEOUT_MS)
+				return -EBUSY;
+		}
+	}
 	return 0;
 }
 
@@ -174,7 +195,33 @@
 
 uint32_t ipm_gpu_bridge_max_id_val_get(const struct device *dev)
 {
-	ARG_UNUSED(dev);
+	const struct ipm_gpu_bridge_config *config = DEV_CFG(dev);
+	struct gpu_bridge_cfg_regs *cfg_base = (struct gpu_bridge_cfg_regs *)(config->cfg_base);
+	struct gpu_bridge_intr_regs *intr_base = (struct gpu_bridge_intr_regs *)(config->intr_base);
+	struct ipm_gpu_bridge_data *const data = DEV_DATA(dev);
+	uint32_t interrupt_data[INTR_MSG_HDR_SZ] = {0};
+	int i;
+	uint32_t size;
+
+	if (!sys_read32((mem_addr_t) &cfg_base->interrupt_out_valid)) {
+		LOG_INF("No interrupt received for bridge dev: %s", dev->name);
+		return 0;
+	}
+	size = (sys_read32((mem_addr_t) &cfg_base->interrupt_out_cfg) &
+				GPU_BRIDGE_NUM_FLITS_MASK) * 4;
+	if (!size) {
+		/* Clear interrupt and return */
+		LOG_INF("Interrupt received with 0 data for bridge dev: %s", dev->name);
+		sys_write32(0x1, (mem_addr_t) &cfg_base->interrupt_out_valid);
+		sys_write32(BIT(config->bridge_idx), (mem_addr_t) &intr_base->cause);
+		return 0;
+	}
+
+	for (i = 0; i < INTR_MSG_HDR_SZ; i++) {
+		interrupt_data[i] = sys_read32((mem_addr_t)&cfg_base->interrupt_out_data[i]);
+	}
+
+	data->callback(dev, data->user_data, size, &interrupt_data);
 
 	return 0;
 }
@@ -206,7 +253,7 @@
 		return 0;
 	}
 
-	sys_write32(0x1, (mem_addr_t) &cfg_base->interrupt_out_valid);
+	sys_write32(0x0, (mem_addr_t) &cfg_base->interrupt_out_valid);
 	sys_write32(BIT(config->bridge_idx), (mem_addr_t) &intr_base->cause);
 
 	return 0;
